contracts/utils/cryptography/RSA.sol

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Math} from "../math/Math.sol";
+
+/**
+ *  TODO:
+ *  - Further optimize ?
+ *  - Write documentation
+ *
+ *  Inspired by Adrià Massanet's work: https://github.com/adria0/SolRsaVerify
+ */
+library RSA {
+    /**
+     * @dev Verifies a PKCSv1.5 SHA256 signature
+     * @param data to verify
+     * @param sig is the signature
+     * @param exp is the exponent
+     * @param mod is the modulus
+     */
+    function pkcs1Sha256(
+        bytes memory data,
+        bytes memory sig,
+        bytes memory exp,
+        bytes memory mod
+    ) internal view returns (bool) {
+        return pkcs1Sha256(sha256(data), sig, exp, mod);
+    }
+
+    /**
+     * @dev Verifies a PKCSv1.5 SHA256 signature
+     * @param digest is the sha256 of the data
+     * @param sig is the signature
+     * @param exp is the exponent
+     * @param mod is the modulus
+     */
+    function pkcs1Sha256(
+        bytes32 digest,
+        bytes memory sig,
+        bytes memory exp,
+        bytes memory mod
+    ) internal view returns (bool) {
+        unchecked {
+            // cache and check length
+            uint256 length = mod.length;
+            if (length < 0x40 || length != sig.length) {
```

Comment:
I guess we reject length less than `0x40` because it wouldn't be secure. I wonder if `0x40` was arbitrarily chosen. If so, we need to evaluate it carefully, as far as I remember, RSA's security is `p * q` so a 512 bits signature is crackable in reasonable time.

Found this as a reference, but seems like 512 bits (0x40 bytes) signatures are pretty much broken.
https://github.com/tomrittervg/cloud-and-control/blob/master/gnfs-info/factoring-howto.txt

[RFC 3447](https://datatracker.ietf.org/doc/html/rfc8017) is from 2003 and was superseded by [RFC 8017](https://datatracker.ietf.org/doc/html/rfc8017), though, I couldn't find a recommendation for the mod length. Allegedly, 512 bits security was [first broken in 1999](https://crypto.stackexchange.com/a/3933), so my estimations say that we might increase this to `0x80` at least.

Needs discussion